src/swapRouter.ts

```diff
@@ -302,14 +330,97 @@ export abstract class SwapRouter {
       }
     }
 
-    // refund
-    if (mustRefund) {
+    // must refund when paying in ETH, but with an uncertain input amount
+    if (inputIsNative && sampleTrade.tradeType === TradeType.EXACT_OUTPUT) {
       calldatas.push(Payments.encodeRefundETH())
     }
 
     return {
       calldata: MulticallExtended.encodeMulticall(calldatas, options.deadlineOrPreviousBlockhash),
-      value: toHex(totalValue.quotient),
+      value: toHex(inputIsNative ? totalAmountIn.quotient : ZERO),
     }
   }
+
+  /**
+   * Produces the on-chain method name to call and the hex encoded parameters to pass as arguments for a given trade.
+   * @param trade to produce call parameters for
+   * @param options options for the call parameters
+   */
+  public static swapAndAddCallParameters(
+    trades:
+      | Trade<Currency, Currency, TradeType>
+      | V2Trade<Currency, Currency, TradeType>
+      | V3Trade<Currency, Currency, TradeType>
+      | (V2Trade<Currency, Currency, TradeType> | V3Trade<Currency, Currency, TradeType>)[],
+    options: SwapOptions,
+    position: Position,
+    addLiquidityOptions: AddLiquidityOptions
+  ): MethodParameters {
+    const {
+      calldatas,
+      inputIsNative,
+      outputIsNative,
+      totalAmountIn: totalAmountSwapped,
+      totalAmountOut,
+    } = SwapRouter.encodeSwaps(trades, options, true)
+
+    const zeroForOne = position.pool.token0 === totalAmountSwapped.currency.wrapped
+    const { positionAmountIn, positionAmountOut } = SwapRouter.getPositionAmounts(position, zeroForOne)
+
+    // if tokens are native they will be converted to WETH9
+    const tokenIn = inputIsNative ? WETH9[1] : positionAmountIn.currency.wrapped
+    const tokenOut = outputIsNative ? WETH9[1] : positionAmountOut.currency.wrapped
+
+    // if swap output does not make up whole outputTokenBalanceDesired, pull in remaining tokens for adding liquidity
+    const amountOutRemaining = positionAmountOut.subtract(totalAmountOut.wrapped)
+    if (amountOutRemaining.greaterThan(CurrencyAmount.fromRawAmount(positionAmountOut.currency, 0))) {
+      // if output is native, convert value to WETH9, else pull ERC20 token
+      outputIsNative
+        ? calldatas.push(SwapRouter.INTERFACE.encodeFunctionData('wrapETH', [amountOutRemaining.quotient]))
```

Comment:
so here i guess we're using native output as a flag for wanting to pay the remainder in ETH, or not for WETH?

Reply:
hmmm....yah. I guess this is a weird situation in which you list your output token as native, even though it will never actually get converted back to native since it's going into the position. But it allows you to send the "remaining" portion as native. If the wording is too indirect I can try and think of something else...lmk what you think.

Reply:
maybe just a brief comment explaining? i think it's fine otherwise